test/lit/passes/signature-refining.wast

```diff
@@ -490,3 +490,137 @@
     )
   )
 )
+
+(module
+  ;; CHECK:      (type $struct (struct_subtype  data))
+  (type $struct (struct_subtype data))
+
+  ;; This signature has a single function using it, which returns a more
+  ;; refined type, and we can refine to that.
+  ;; CHECK:      (type $sig-can-refine (func_subtype (result (ref $struct)) func))
+  (type $sig-can-refine (func_subtype (result anyref) func))
+
+  ;; Also a single function, but no refinement is possible.
+  ;; CHECK:      (type $sig-cannot-refine (func_subtype (result anyref) func))
+  (type $sig-cannot-refine (func_subtype (result anyref) func))
+
+  ;; The single function never returns, so no refinement is possible.
+  ;; CHECK:      (type $sig-unreachable (func_subtype (result anyref) func))
+  (type $sig-unreachable (func_subtype (result anyref) func))
+
+  ;; CHECK:      (type $none_=>_none (func_subtype func))
+
+  ;; CHECK:      (elem declare func $func-can-refine)
+
+  ;; CHECK:      (func $func-can-refine (type $sig-can-refine) (result (ref $struct))
+  ;; CHECK-NEXT:  (struct.new_default $struct)
+  ;; CHECK-NEXT: )
+  (func $func-can-refine (type $sig-can-refine) (result anyref)
```

Comment:
In the new nominal type syntax, do we write both `type` and `result` in a function signature? Doesn't `type` include `result`? Why do we need to specify `result` redundantly?

Reply:
Yes, the type name does imply the types of the params and the results, so there is redundancy here. But two functions with the same params and results can now have different nominal types (that is, the function's types are structurally the same, but not nominally). So we need to write out the type name. We could perhaps not write out the params and results in that case, but I'm not sure the wat format allows such shorthand? Even if it does though it seems clearer to write it out I think (so people don't need to scroll up to find the type's definition).